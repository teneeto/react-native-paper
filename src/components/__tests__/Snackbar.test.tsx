import * as React from 'react';
import { Animated, StyleSheet, Text, View } from 'react-native';

import { render } from '@testing-library/react-native';
import renderer from 'react-test-renderer';

import { red200, white } from '../../styles/themes/v2/colors';
import Snackbar from '../Snackbar';

const styles = StyleSheet.create({
  snackContent: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  iconView: {
    backgroundColor: red200,
    padding: 15,
  },
  text: { color: white, marginLeft: 10, flexWrap: 'wrap', flexShrink: 1 },
});

// Make sure any animation finishes before checking the snapshot results
jest.mock('react-native', () => {
  const RN = jest.requireActual('react-native');

  const timing: typeof Animated['timing'] = (value, config) => ({
    start: (callback) => {
      value.setValue(config.toValue as any);
      callback?.({ finished: true });
    },
    value,
    config,
    stop: () => {
      throw new Error('Not implemented');
    },
    reset: () => {
      throw new Error('Not implemented');
    },
  });
  RN.Animated.timing = timing;

  return RN;
});

jest.mock('react-native-safe-area-context', () => ({
  useSafeAreaInsets: () => ({ bottom: 34, left: 0, right: 0, top: 47 }),
}));

it('renders snackbar with content', () => {
  const tree = renderer
    .create(
      <Snackbar visible onDismiss={jest.fn()}>
        Snackbar content
      </Snackbar>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders not visible snackbar with content wrapper but no actual content', () => {
  const tree = renderer
    .create(
      <Snackbar visible={false} onDismiss={jest.fn()}>
        Snackbar content
      </Snackbar>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders snackbar with Text as a child', () => {
  const tree = renderer
    .create(
      <Snackbar visible onDismiss={jest.fn()}>
        <Text>Snackbar content</Text>
      </Snackbar>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders snackbar with action button', () => {
  const tree = renderer
    .create(
      <Snackbar
        visible
        onDismiss={() => {}}
        action={{ label: 'Undo', onPress: jest.fn() }}
      >
        Snackbar content
      </Snackbar>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('renders snackbar with View & Text as a child', () => {
  const tree = renderer
    .create(
      <Snackbar visible onDismiss={jest.fn()}>
        <View style={styles.snackContent}>
          <View style={styles.iconView} />
          <Text style={styles.text}>
            Error Message which is veryyyyyyyyyyyy longggggggg Error Message
            which is veryyyyyyyyyyyy longggggggg
          </Text>
        </View>
      </Snackbar>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('animated value changes correctly', () => {
  const value = new Animated.Value(1);
  const { getByTestId } = render(
    <Snackbar
      visible
      onDismiss={jest.fn()}
      testID="snack-bar"
      style={[{ transform: [{ scale: value }] }]}
    >
      Snackbar content
    </Snackbar>
  );
  expect(getByTestId('snack-bar-outer-layer')).toHaveStyle({
    transform: [{ scale: 1 }],
  });

  Animated.timing(value, {
    toValue: 1.5,
    useNativeDriver: false,
    duration: 200,
  }).start();

  jest.advanceTimersByTime(200);

  expect(getByTestId('snack-bar-outer-layer')).toHaveStyle({
    transform: [{ scale: 1.5 }],
  });
});
