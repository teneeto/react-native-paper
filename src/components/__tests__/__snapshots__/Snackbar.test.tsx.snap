// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders not visible snackbar with content wrapper but no actual content 1`] = `null`;

exports[`renders snackbar with Text as a child 1`] = `
<View
  pointerEvents="box-none"
  style={
    Array [
      Object {
        "bottom": 0,
        "position": "absolute",
        "width": "100%",
      },
      Object {
        "paddingBottom": 34,
        "paddingHorizontal": 0,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    style={
      Object {
        "alignSelf": undefined,
        "backgroundColor": "rgba(49, 48, 51, 1)",
        "borderRadius": 4,
        "bottom": undefined,
        "end": undefined,
        "flex": undefined,
        "height": undefined,
        "left": undefined,
        "margin": 8,
        "opacity": 1,
        "position": undefined,
        "right": undefined,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.15,
        "shadowRadius": 6,
        "start": undefined,
        "top": undefined,
        "transform": Array [
          Object {
            "scale": 1,
          },
        ],
        "width": undefined,
      }
    }
    testID="surface-outer-layer"
  >
    <View
      accessibilityLiveRegion="polite"
      collapsable={false}
      pointerEvents="box-none"
      style={
        Object {
          "backgroundColor": "rgba(49, 48, 51, 1)",
          "borderRadius": 4,
          "flex": undefined,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "minHeight": 48,
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 1,
            "width": 0,
          },
          "shadowOpacity": 0.3,
          "shadowRadius": 2,
        }
      }
      testID="surface"
    >
      <View
        style={
          Object {
            "flex": 1,
            "marginHorizontal": 16,
            "marginVertical": 14,
          }
        }
      >
        <View>
          <Text>
            Snackbar content
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders snackbar with View & Text as a child 1`] = `
<View
  pointerEvents="box-none"
  style={
    Array [
      Object {
        "bottom": 0,
        "position": "absolute",
        "width": "100%",
      },
      Object {
        "paddingBottom": 34,
        "paddingHorizontal": 0,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    style={
      Object {
        "alignSelf": undefined,
        "backgroundColor": "rgba(49, 48, 51, 1)",
        "borderRadius": 4,
        "bottom": undefined,
        "end": undefined,
        "flex": undefined,
        "height": undefined,
        "left": undefined,
        "margin": 8,
        "opacity": 1,
        "position": undefined,
        "right": undefined,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.15,
        "shadowRadius": 6,
        "start": undefined,
        "top": undefined,
        "transform": Array [
          Object {
            "scale": 1,
          },
        ],
        "width": undefined,
      }
    }
    testID="surface-outer-layer"
  >
    <View
      accessibilityLiveRegion="polite"
      collapsable={false}
      pointerEvents="box-none"
      style={
        Object {
          "backgroundColor": "rgba(49, 48, 51, 1)",
          "borderRadius": 4,
          "flex": undefined,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "minHeight": 48,
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 1,
            "width": 0,
          },
          "shadowOpacity": 0.3,
          "shadowRadius": 2,
        }
      }
      testID="surface"
    >
      <View
        style={
          Object {
            "flex": 1,
            "marginHorizontal": 16,
            "marginVertical": 14,
          }
        }
      >
        <View>
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
              }
            }
          >
            <View
              style={
                Object {
                  "backgroundColor": "#ef9a9a",
                  "padding": 15,
                }
              }
            />
            <Text
              style={
                Object {
                  "color": "#ffffff",
                  "flexShrink": 1,
                  "flexWrap": "wrap",
                  "marginLeft": 10,
                }
              }
            >
              Error Message which is veryyyyyyyyyyyy longggggggg Error Message which is veryyyyyyyyyyyy longggggggg
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders snackbar with action button 1`] = `
<View
  pointerEvents="box-none"
  style={
    Array [
      Object {
        "bottom": 0,
        "position": "absolute",
        "width": "100%",
      },
      Object {
        "paddingBottom": 34,
        "paddingHorizontal": 0,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    style={
      Object {
        "alignSelf": undefined,
        "backgroundColor": "rgba(49, 48, 51, 1)",
        "borderRadius": 4,
        "bottom": undefined,
        "end": undefined,
        "flex": undefined,
        "height": undefined,
        "left": undefined,
        "margin": 8,
        "opacity": 1,
        "position": undefined,
        "right": undefined,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.15,
        "shadowRadius": 6,
        "start": undefined,
        "top": undefined,
        "transform": Array [
          Object {
            "scale": 1,
          },
        ],
        "width": undefined,
      }
    }
    testID="surface-outer-layer"
  >
    <View
      accessibilityLiveRegion="polite"
      collapsable={false}
      pointerEvents="box-none"
      style={
        Object {
          "backgroundColor": "rgba(49, 48, 51, 1)",
          "borderRadius": 4,
          "flex": undefined,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "minHeight": 48,
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 1,
            "width": 0,
          },
          "shadowOpacity": 0.3,
          "shadowRadius": 2,
        }
      }
      testID="surface"
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "System",
              "fontSize": 14,
              "fontWeight": "400",
              "letterSpacing": 0.25,
              "lineHeight": 20,
            },
            Object {
              "textAlign": "left",
            },
            Object {
              "color": "rgba(28, 27, 31, 1)",
              "writingDirection": "ltr",
            },
            Array [
              Object {
                "flex": 1,
                "marginHorizontal": 16,
                "marginVertical": 14,
              },
              Object {
                "color": "rgba(244, 239, 244, 1)",
              },
            ],
          ]
        }
      >
        Snackbar content
      </Text>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "flex-end",
              "minHeight": 48,
            },
            Object {
              "marginLeft": -12,
            },
          ]
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "alignSelf": undefined,
              "backgroundColor": "transparent",
              "borderRadius": 20,
              "bottom": undefined,
              "end": undefined,
              "flex": undefined,
              "height": undefined,
              "left": undefined,
              "marginLeft": 4,
              "marginRight": 8,
              "opacity": undefined,
              "position": undefined,
              "right": undefined,
              "shadowColor": "#000",
              "shadowOffset": Object {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
              "start": undefined,
              "top": undefined,
              "transform": undefined,
              "width": undefined,
            }
          }
          testID="button-container-outer-layer"
        >
          <View
            collapsable={false}
            style={
              Object {
                "backgroundColor": "transparent",
                "borderColor": "transparent",
                "borderRadius": 20,
                "borderStyle": "solid",
                "borderWidth": 0,
                "flex": undefined,
                "minWidth": 64,
                "shadowColor": "#000",
                "shadowOffset": Object {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 0,
                "shadowRadius": 0,
              }
            }
            testID="button-container"
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                Object {
                  "disabled": false,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Array [
                  Object {
                    "overflow": "hidden",
                  },
                  false,
                  Object {
                    "borderRadius": 20,
                  },
                ]
              }
              testID="button"
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "center",
                    },
                    undefined,
                  ]
                }
              >
                <Text
                  numberOfLines={1}
                  selectable={false}
                  style={
                    Array [
                      Object {
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0.1,
                        "lineHeight": 20,
                      },
                      Object {
                        "textAlign": "left",
                      },
                      Object {
                        "color": "rgba(28, 27, 31, 1)",
                        "writingDirection": "ltr",
                      },
                      Array [
                        Object {
                          "marginHorizontal": 16,
                          "marginVertical": 9,
                          "textAlign": "center",
                        },
                        false,
                        Object {
                          "marginHorizontal": 12,
                        },
                        false,
                        false,
                        Object {
                          "color": "rgba(208, 188, 255, 1)",
                          "fontFamily": "System",
                          "fontSize": 14,
                          "fontWeight": "500",
                          "letterSpacing": 0.1,
                          "lineHeight": 20,
                        },
                        undefined,
                      ],
                    ]
                  }
                  testID="button-text"
                >
                  Undo
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders snackbar with content 1`] = `
<View
  pointerEvents="box-none"
  style={
    Array [
      Object {
        "bottom": 0,
        "position": "absolute",
        "width": "100%",
      },
      Object {
        "paddingBottom": 34,
        "paddingHorizontal": 0,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    style={
      Object {
        "alignSelf": undefined,
        "backgroundColor": "rgba(49, 48, 51, 1)",
        "borderRadius": 4,
        "bottom": undefined,
        "end": undefined,
        "flex": undefined,
        "height": undefined,
        "left": undefined,
        "margin": 8,
        "opacity": 1,
        "position": undefined,
        "right": undefined,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.15,
        "shadowRadius": 6,
        "start": undefined,
        "top": undefined,
        "transform": Array [
          Object {
            "scale": 1,
          },
        ],
        "width": undefined,
      }
    }
    testID="surface-outer-layer"
  >
    <View
      accessibilityLiveRegion="polite"
      collapsable={false}
      pointerEvents="box-none"
      style={
        Object {
          "backgroundColor": "rgba(49, 48, 51, 1)",
          "borderRadius": 4,
          "flex": undefined,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "minHeight": 48,
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 1,
            "width": 0,
          },
          "shadowOpacity": 0.3,
          "shadowRadius": 2,
        }
      }
      testID="surface"
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "System",
              "fontSize": 14,
              "fontWeight": "400",
              "letterSpacing": 0.25,
              "lineHeight": 20,
            },
            Object {
              "textAlign": "left",
            },
            Object {
              "color": "rgba(28, 27, 31, 1)",
              "writingDirection": "ltr",
            },
            Array [
              Object {
                "flex": 1,
                "marginHorizontal": 16,
                "marginVertical": 14,
              },
              Object {
                "color": "rgba(244, 239, 244, 1)",
              },
            ],
          ]
        }
      >
        Snackbar content
      </Text>
    </View>
  </View>
</View>
`;
